File: abgui/Sources/abgui/Models/CSVDocument.swift
// Copyright 2026 Gigaion, LLC
// SPDX-License-Identifier: AGPL-3.0-or-later
import SwiftUI
import UniformTypeIdentifiers

/// An RFC-4180 CSV file for `.fileExporter` (the read-only list screens' Export CSV).
/// Mirrors abctl's CSV output (internal/cli/output.go): fields containing a comma,
/// quote, or newline are quoted with embedded quotes doubled, and tenant-controlled
/// cell values are neutralized against spreadsheet formula injection.
struct CSVDocument: FileDocument {
    static var readableContentTypes: [UTType] { [.commaSeparatedText] }

    /// The finished CSV text (header row first, LF-terminated lines like abctl's).
    let text: String

    /// Builds the document from column headers + row cells. Cells are formula-
    /// neutralized; headers are our own literals, so they are quoted but not prefixed.
    init(headers: [String], rows: [[String]]) {
        var lines = [Self.line(headers)]
        for row in rows {
            lines.append(Self.line(row.map(Self.sanitize)))
        }
        text = lines.joined(separator: "\n") + "\n"
    }

    init(configuration: ReadConfiguration) throws {
        guard let data = configuration.file.regularFileContents,
              let string = String(data: data, encoding: .utf8) else {
            throw CocoaError(.fileReadCorruptFile)
        }
        text = string
    }

    func fileWrapper(configuration: WriteConfiguration) throws -> FileWrapper {
        FileWrapper(regularFileWithContents: Data(text.utf8))
    }

    private static func line(_ fields: [String]) -> String {
        fields.map(quote).joined(separator: ",")
    }

    /// RFC-4180 quoting: wrap the field in quotes when it contains a comma, quote,
    /// or newline, doubling any embedded quotes; otherwise emit it verbatim.
    private static func quote(_ field: String) -> String {
        guard field.contains(",") || field.contains("\"")
                || field.contains("\n") || field.contains("\r") else { return field }
        return "\"" + field.replacingOccurrences(of: "\"", with: "\"\"") + "\""
    }

    /// Neutralizes spreadsheet formula injection exactly like abctl's csvSanitize:
    /// cells starting with '=', '+', '-', '@', tab, or CR are interpreted as formulas
    /// by Excel/LibreOffice/Google Sheets, so they get a leading single quote — the
    /// standard mitigation, rendered as a literal by spreadsheets.
    private static func sanitize(_ field: String) -> String {
        guard let first = field.first else { return field }
        switch first {
        case "=", "+", "-", "@", "\t", "\r":
            return "'" + field
        default:
            return field
        }
    }
}

File: abgui-flutter/lib/src/models/csv_document.dart
// Copyright 2026 Gigaion, LLC
// SPDX-License-Identifier: AGPL-3.0-or-later

// Port note (rule 6): the Swift type conformed to SwiftUI's `FileDocument` so `.fileExporter`
// could hand it to a save panel. That conformance is macOS plumbing, not contract — in Flutter
// the save dialog comes from `file_selector` and wants bytes — so what survives here is the
// part that IS the contract: the encoding. It is a pure function of headers and rows, which
// also makes it testable without a widget tree or a file system.

/// Builds an RFC-4180 CSV document (the read-only list screens' Export CSV).
///
/// Mirrors abctl's CSV output (internal/cli/output.go): fields containing a comma, quote, or
/// newline are quoted with embedded quotes doubled, and tenant-controlled cell values are
/// neutralized against spreadsheet formula injection. Header row first, LF-terminated lines
/// like abctl's — so a CSV exported from the GUI diffs cleanly against one piped from the CLI.
///
/// Cells are formula-neutralized; headers are our own literals, so they are quoted but not
/// prefixed.
String csvDocument({
  required List<String> headers,
  required List<List<String>> rows,
}) {
  final lines = <String>[_csvLine(headers)];
  for (final row in rows) {
    lines.add(_csvLine(row.map(csvSanitizeCell).toList(growable: false)));
  }
  return '${lines.join('\n')}\n';
}

String _csvLine(List<String> fields) => fields.map(csvQuoteField).join(',');

/// RFC-4180 quoting: wrap the field in quotes when it contains a comma, quote, or newline,
/// doubling any embedded quotes; otherwise emit it verbatim.
String csvQuoteField(String field) {
  if (!field.contains(',') &&
      !field.contains('"') &&
      !field.contains('\n') &&
      !field.contains('\r')) {
    return field;
  }
  return '"${field.replaceAll('"', '""')}"';
}

/// Neutralizes spreadsheet formula injection exactly like abctl's csvSanitize: cells starting
/// with '=', '+', '-', '@', tab, or CR are interpreted as formulas by Excel/LibreOffice/Google
/// Sheets, so they get a leading single quote — the standard mitigation, rendered as a literal
/// by spreadsheets.
///
/// Note the interaction with [csvQuoteField], which runs afterwards: a cell starting with CR
/// becomes `'\r…`, which then quotes as a multi-line field. That ordering is abctl's and is
/// preserved deliberately — sanitize decides the CONTENT, quote decides the FRAMING.
String csvSanitizeCell(String field) {
  if (field.isEmpty) return field;
  switch (field[0]) {
    case '=':
    case '+':
    case '-':
    case '@':
    case '\t':
    case '\r':
      return "'$field";
    default:
      return field;
  }
}
